File: services/app/domain/magnitude.py
"""How unusual is this move, in units of the stock's own volatility?

Spec 3.2:   magnitude = |r| / (sigma * sqrt(N))

This module is the Python reference implementation. The Angular client computes the
same thing (spec 6.3), and both are pinned to one shared golden fixture so the two
halves of the formula cannot drift apart across the language boundary.
"""
from __future__ import annotations

import math
from enum import StrEnum

from app.config import BAND_EDGES, CORPORATE_ACTION_RETURN


class Band(StrEnum):
    NOISE = "noise"
    MILD = "mild"
    NOTABLE = "notable"
    EXTREME = "extreme"
    NEW = "new"                          # spec 4.7: no baseline to compare against
    UNRATED = "unrated"                  # spec 4.7: the stock has no sigma yet
    CORPORATE_ACTION = "corporate_action" # spec 9: flagged, never ranked


def simple_return(baseline: float, observed: float) -> float:
    return (observed - baseline) / baseline


def magnitude(baseline: float | None, observed: float | None,
              sigma: float | None, n: int) -> float | None:
    """None whenever the inputs cannot support a ranking, never a faked default.

    Spec 4.7 is honest that the ranking layer is where the two NULLs stop being
    free: storage treats them uniformly, arithmetic cannot.
    """
    if baseline is None or observed is None or sigma is None:
        return None
    if baseline <= 0.0 or sigma <= 0.0 or n <= 0:
        return None
    expected = sigma * math.sqrt(n)
    return abs(simple_return(baseline, observed)) / expected


def is_corporate_action(r: float | None) -> bool:
    """Keyed off the return, which does not move with N.

    An earlier draft tested `magnitude > 10`. Magnitude scales with sqrt(N), so a
    user logging in twice within a few minutes would see ordinary moves labelled
    corporate actions - and banding would depend on N, contradicting spec 3.5.
    """
    return r is not None and abs(r) >= CORPORATE_ACTION_RETURN


def band(m: float | None, *, has_baseline: bool, has_sigma: bool,
         r: float | None = None) -> Band:
    """Half-open intervals, so every boundary value lands in exactly one band."""
    if not has_baseline:
        return Band.NEW
    if is_corporate_action(r):
        # A 1:10 split is -90% and would otherwise top every watchlist forever.
        return Band.CORPORATE_ACTION
    if not has_sigma:
        return Band.UNRATED
    if m is None:
        return Band.UNRATED
    mild, notable, extreme = BAND_EDGES
    if m >= extreme:
        return Band.EXTREME
    if m >= notable:
        return Band.NOTABLE
    if m >= mild:
        return Band.MILD
    return Band.NOISE


def annualised(sigma_per_bucket: float, buckets_per_day: int, trading_days: int = 250) -> float:
    """Sanity check, not decoration.

    Any sigma we quote has to survive this conversion; figures that annualise to
    150%+ mean the example was arithmetic rather than a market (spec 3.3).
    """
    return sigma_per_bucket * math.sqrt(buckets_per_day * trading_days)

File: web/src/app/domain/magnitude.ts
/**
 * magnitude = |r| / (sigma * sqrt(N))        - design document, section 3.2
 *
 * This is the compute layer, not the view layer: the server performs no per-user
 * arithmetic at all (section 6.3), so this file is where the product's thesis
 * actually executes. It is pure, imports nothing from Angular, and is pinned to the
 * same worked example as the Python reference so the two cannot drift.
 */

export const BAND_EDGES = { mild: 1, notable: 2, extreme: 3 };
/**
 * Keyed off the RETURN, not the magnitude. Magnitude scales with sqrt(N), so a
 * magnitude threshold made banding depend on N - an ordinary +2.7% move read as a
 * corporate action whenever a user logged in twice within a few minutes, which also
 * contradicted section 3.5. A 1:10 split is -90%, a 1:2 split is -50%, and an NSE
 * circuit band caps a real session move at 20%.
 */
export const CORPORATE_ACTION_RETURN = 0.4;

export function isCorporateAction(r: number | null): boolean {
  return r !== null && Math.abs(r) >= CORPORATE_ACTION_RETURN;
}

export function simpleReturn(baseline: number, observed: number): number {
  return (observed - baseline) / baseline;
}

export function magnitude(
  baseline: number | null,
  observed: number | null,
  sigma: number | null,
  n: number,
): number | null {
  // None whenever the inputs cannot support a ranking - never a faked default.
  if (baseline === null || observed === null || sigma === null) return null;
  if (baseline <= 0 || sigma <= 0 || n <= 0) return null;
  const value = Math.abs(simpleReturn(baseline, observed)) / (sigma * Math.sqrt(n));
  return Number.isFinite(value) ? value : null;
}

/**
 * Half-open intervals, so every boundary value lands in exactly one band.
 * A magnitude of exactly 2 is `notable`, never ambiguous.
 */
export function band(
  m: number | null,
  opts: { hasBaseline: boolean; hasSigma: boolean; atCircuit?: boolean; r?: number | null },
): string {
  if (!opts.hasBaseline) return 'new';
  if (isCorporateAction(opts.r ?? null)) return 'corporate_action';
  if (opts.atCircuit) return 'at_circuit';
  if (!opts.hasSigma || m === null) return 'unrated';
  if (m >= BAND_EDGES.extreme) return 'extreme';
  if (m >= BAND_EDGES.notable) return 'notable';
  if (m >= BAND_EDGES.mild) return 'mild';
  return 'noise';
}

/**
 * Is the price pinned at its official circuit band?
 *
 * The band is published against the previous close; we compare against the session
 * baseline, which is the closest thing the client holds. Good to within a day's
 * drift, and the approximation is stated rather than hidden.
 */
export function atCircuit(
  bandLimit: string, baseline: number | null, price: number | null,
): boolean {
  if (baseline === null || price === null) return false;
  const match = /^(\d+(?:\.\d+)?)%$/.exec(bandLimit.trim());
  if (!match) return false;                       // "No Band" - every F&O stock
  const limit = parseFloat(match[1]) / 100;
  return Math.abs(simpleReturn(baseline, price)) >= limit * 0.99;
}
